File: typescript/src/ne/restable.ts
import ByteStream from "../bytestream";

export enum ResourceTypes {
    Cursor = 0x8001,
    Bitmap,
    Icon,
    Menu,
    Dialog,
    String,
    FontDir,
    Font,
    Accelerator,
    RCData,
    MessageTable,
    GroupCursor,
    GroupIcon = 0x800E,
    Version = 0x8010,
    DLGInclude,
    PlugPlay = 0x8013,
    VXD,
    AniCursor,
    AniIcon,
    HTML,
    Manifest,
}

export class ResourceTable {
    AlignShift: number;
    Types: TypeInfo[];
    ResourceStrings: string[];

    private constructor(alignShift: number, types: TypeInfo[], resStrs: string[]) {
        this.AlignShift = alignShift;
        this.Types = types;
        this.ResourceStrings = resStrs;
    }

    public static Read(bst: ByteStream, length: number) {
        const _bstStart = bst.Position;

        const alignShift = bst.ReadWord();

        const types: TypeInfo[] = [];
        while (bst.PeekWord() != 0) {
            types.push(TypeInfo.Read(bst));
        }
        bst.ReadWord(); // garbage

        // alright, this is weird.  The "if" check is literally just to make sure that the resource table doesn't actually have any strings.
        // It shows in every valid doc I've seen that it should exist, but the exe's I've been using to test this, it doesn't exist.
        if (bst.Position - _bstStart < length) {
            let resStrs: string[] = [];
            while (bst.PeekByte() != 0) {
                resStrs.push(bst.ReadString(bst.ReadByte()));
            }
            bst.ReadByte(); // also garbage
        }

        return new ResourceTable(alignShift, types, []);
    }
}

export class TypeInfo {
    TypeIDOrOffset: number;
    ResourceCount: number;
    NameInfo: NameInfo[];

    private constructor(typeId: number, resCount: number, nameInfo: NameInfo[]) {
        this.TypeIDOrOffset = typeId;
        this.ResourceCount = resCount;
        this.NameInfo = nameInfo;
    }

    public static Read(bst: ByteStream) {
        const typeid = bst.ReadWord();
        const nRes = bst.ReadWord();
        if (!bst.CheckReserved(4)) {
            debugger;
        }

        const names: NameInfo[] = [];
        for (let i = 0; i < nRes; i++) {
            names.push(NameInfo.Read(bst));
        }

        return new TypeInfo(typeid, nRes, names);
    }
}

export class NameInfo {
    Offset: number;
    Length: number;
    Flags: number;
    ID: number;

    private constructor(offset: number, len: number, flags: number, id: number) {
        this.Offset = offset;
        this.Length = len;
        this.Flags = flags;
        this.ID = id;
    }

    public static Read(bst: ByteStream) {
        const offset = bst.ReadWord();
        const length = bst.ReadWord();
        const flag = bst.ReadWord();
        const resid = bst.ReadWord();
        if (!bst.CheckReserved(4)) {
            debugger;
        }

        return new NameInfo(offset, length, flag, resid);
    }
}

File: rust/src/executable/mz/ne/restable.rs
use crate::byte_stream::ByteStream;

pub enum ResourceTypes {
    Cursor = 0x8001,
    Bitmap,
    Icon,
    Menu,
    Dialog,
    String,
    FontDir,
    Font,
    Accelerator,
    RCData,
    MessageTable,
    GroupCursor,
    GroupIcon = 0x800E,
    Version = 0x8010,
    DLGInclude,
    PlugPlay = 0x8013,
    VXD,
    AniCursor,
    AniIcon,
    HTML,
    Manifest,
}

pub struct ResourceTable {
    pub align_shift: u16,
    pub types: Vec<TypeInfo>,
    pub resource_strings: Vec<String>,
}
impl ResourceTable {
    pub fn read(bst: &mut ByteStream, length: u16) -> Self {
        let bst_start = bst.pos;

        let align_shift = bst.read_word();
        let mut types = Vec::new();
        while bst.peek_word() != 0 {
            types.push(TypeInfo::read(bst));
        }
        bst.read_word();

        let mut resource_strings = Vec::new();
        if bst.pos - bst_start < length as usize {
            while bst.peek_byte() != 0 {
                let _t = bst.read_byte();
                resource_strings.push(bst.read_string(_t as usize));
            }
            _ = bst.read_byte();
        }

        Self {
            align_shift,
            types,
            resource_strings,
        }
    }
}

pub struct TypeInfo {
    pub type_id_or_offset: u16,
    pub res_count: u16,
    pub name_info: Vec<NameInfo>,
}
impl TypeInfo {
    pub fn read(bst: &mut ByteStream) -> Self {
        let type_id_or_offset = bst.read_word();
        let res_count = bst.read_word();
        bst.pos += 4; // bst.check_reserved(4);
        let mut name_info = Vec::new();
        for _ in 0..res_count {
            name_info.push(NameInfo::read(bst));
        }
        Self {
            type_id_or_offset,
            res_count,
            name_info,
        }
    }
}

pub struct NameInfo {
    pub offset: u16,
    pub length: u16,
    pub flags: u16,
    pub id: u16,
}
impl NameInfo {
    pub fn read(bst: &mut ByteStream) -> Self {
        let offset = bst.read_word();
        let length = bst.read_word();
        let flags = bst.read_word();
        let id = bst.read_word();
        bst.pos += 4; // bst.check_reserved(4);
        Self {
            offset,
            length,
            flags,
            id,
        }
    }
}
